File: packages/react-native/android/src/main/kotlin/com/uulab/securekeypad/SecureKeypadPresentation.kt
package com.uulab.securekeypad

internal const val SECURE_KEYPAD_MAX_RENDERED_LENGTH = 4_096
internal const val SECURE_KEYPAD_ERROR_INTERNAL = 7

/** Returns masked presentation text without ever accepting a user-supplied value. */
internal fun secureKeypadMaskedDisplayText(length: Int, protected: Boolean = false): String {
    require(length in 0..SECURE_KEYPAD_MAX_RENDERED_LENGTH)
    return if (protected) "Protected" else "•".repeat(length)
}

/** Returns an accessibility announcement containing only state and length. */
internal fun secureKeypadAccessibilityLabel(length: Int, protected: Boolean = false): String {
    require(length in 0..SECURE_KEYPAD_MAX_RENDERED_LENGTH)
    if (protected) return "Protected"
    return if (length == 0) "No input" else "$length characters entered"
}

File: native/ios/SecureKeypadPresentation.swift
import Foundation

let secureKeypadMaxRenderedLength = 4_096
let secureKeypadInternalError: UInt32 = 7

/// Returns whether the native presentation must hide sensitive UI state.
/// Screen capture remains protected across app-active transitions.
func secureKeypadShouldProtectPresentation(applicationIsActive: Bool, screenIsCaptured: Bool) -> Bool {
    !applicationIsActive || screenIsCaptured
}

/// Returns whether a native masked length is safe to render.
func secureKeypadIsValidRenderedLength(_ length: Int) -> Bool {
    (0...secureKeypadMaxRenderedLength).contains(length)
}

/// Produces masked display text from length only.
func secureKeypadMaskedDisplayText(length: Int, protected: Bool) -> String {
    precondition(secureKeypadIsValidRenderedLength(length))
    return protected ? "Protected" : String(repeating: "•", count: length)
}

/// Produces accessibility text from masked state only.
func secureKeypadAccessibilityLabel(length: Int, protected: Bool) -> String {
    precondition(secureKeypadIsValidRenderedLength(length))
    if protected {
        return "Protected"
    }
    return length == 0 ? "No input" : "\(length) characters entered"
}
